enhancements/machine-api/control-plane-machinesets.md

```diff
@@ -0,0 +1,227 @@
+
+- The Installer will create a Machine Health Checking resource targeting the Control Plane Machines.
+
+```
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  name: controlPlane 
+  namespace: openshift-machine-api
+spec:
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: master
+      machine.openshift.io/cluster-api-machine-type: master
+  unhealthyConditions:
+  - type:    "Ready"
+    timeout: "300s"
+    status: "False"
+  - type:    "Ready"
+    timeout: "300s"
+    status: "Unknown"
+  maxUnhealthy: "34%"
+```
+
+-   Any machine deletion will always honour Pod Disruption Budgets (PBD). This gives etcd guard the chance to block a deletion that it considers disruptive.
+
+TODO: Currently multiple MHC can target the same Machine. We should revisit a stronger mechanisim to prevent undesired MHC to target Control Plane Machines.
+
+### Risks and Mitigations
+
+During horizontal scaling operations there are multiple sensitive scenarios like scaling from 1 to 2. As soon as the etcd API is notified of the new member the cluster loses quorum until that new member starts and joins the cluster. This as well as any other operational aspect of etcd healthiness must be still handled by the [Cluster etcd Operator](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md#motivation).
+The contract for the machine API is by honouring Pod Disruption Budgets (PBD), including those given by etcd guard.
+
+## Design Details
+
+### Test Plan
+
+- Exhaustive unit testing.
+- Exhaustive integration testing via [envTest](https://book.kubebuilder.io/reference/testing/envtest.html).
+- E2e testing on the [machine API](https://github.com/openshift/cluster-api-actuator-pkg/tree/master/pkg) and origin e2e test suite:
+	- Scale out.
+	- Scale in.
+	- Vertical scaling forcing machine recreation.
+	- Simulate unhealthy nodes and remediation.
+- E2e testing on the origin e2e test suite.
+
+### Graduation Criteria
+
+All the resources leveraged in this proposal i.e MachineSet and Machine Health Check are already GA beta.
+This proposal will be released in 4.N as long as:
+- All the testing above is in place.
+- The cluster etcd operator manages all the etcd operational aspects including scaling down.
+- There is considerable internal usage and manual disruptive QE testing feedback.
+- This is beta tested pre-release during a considerable period of time either internally or by an early adopter user through the dev preview builds.
+
+### Upgrade / Downgrade Strategy
+
+New IPI clusters deployed after the targeted release will run the MachineSets and MHC deployed by the installer out of the box.
+
+For UPI clusters and existing IPI clusters this is opt-in. We should pursue cluster homogenous topology. To that end we should provide the exact steps and encourage users in any user facing doc to create the MachineSets and MHC for the Control Plane Machines.
```

Comment:
If the machineset and MHC are setup by an operator post-install as recommended in other comments then existing UPI and IPI clusters could adopt this automatically.  Any reason why this wouldn't be true if it's not a part of the installer?

Reply:
Resolving to keep this discussion in one single place https://github.com/openshift/enhancements/pull/292/files#diff-a03db068e1ebe0251c7f4c2e15b5e03fR137